### PR TITLE
some more indexing options (mix of basic, advanced, lists in place of tensors for advanced)

### LINKED ISSUE #643: Indexing with slice and list
Repro:

```
def fn(input_ids):
   return input_ids[:, [-1, 0]]

a = torch.randn(5, 5)
jfn = thunder.jit(fn)
jfn(a)
```

Needed for transformers BERT:

```
import thunder, torch, transformers

m = transformers.BertForSequenceClassification(transformers.BertConfig())
x = torch.randint(1, 10, (1, 32))
jm = thunder.jit(m)
jm(x)
```

Might ideally be fixed along with #460 and #187 

cc @apaz-cli @tfogal

### PR DESCRIPTION
**Pull request by k223kim**
<details>
  <summary><b>Before submitting</b></summary>

- [ ] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [ ] Did you read the [contributor guideline](https://github.com/Lightning-AI/pytorch-lightning/blob/main/.github/CONTRIBUTING.md), Pull Request section?
- [ ] Did you make sure to update the docs?
- [ ] Did you write any new necessary tests?

</details>

## What does this PR do?

Fixes #643, #460, #187.

## PR review

Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?

⚡️

### CODE REVIEW

File: thunder/clang/__init__.py
Comment (mruberry) [inline thunder/clang/__init__.py:730]: Why do you want to convert a list (or other sequence?) to a tensor?

**Comment by mruberry** [comment]
Hey @k223kim!

What's your thinking on the overall strategy to support more (all?) advanced integer indexing?

My understanding may be outdated, but currently I think thunder supports all basic indexing cases, some advanced integer indexing cases, and no advanced boolean indexing cases. Supporting advanced boolean indexing is a whole issue because it involves data-dependent control-flow.

The reason we only support some advanced integer indexing cases is because we couldn't think of an elegant way to map the additional cases to existing thunder operations. One option for supporting more of these cases would be to write a meta function for them, but then just call torch's advanced indexing directly (adding a new primitive for it). The backward for that primitive would likely need to be another primitive that called torch's advanced indexing backward operation directly, too. I'm not sure how easy those operations are to access.

I wonder if we should generally try falling back to the torch operation, and look to devise more special cases where we can express advanced indexing logic by decomposing it to existing operations? Or do you think that we can decompose all advanced indexing cases to existing operations in a reasonable way?

cc @nikitaved

File: thunder/clang/__init__.py
Comment (k223kim) [reply thunder/clang/__init__.py:730]: The reason behind this was because, before this PR, advanced indexing supported Tensor indexing. I thought it would be better to use the existing functionality when receiving a list.

**Comment by k223kim** [comment]
Hi @mruberry ! It's been a while since I got a review from you :) great to see you! Thank you for the detailed feedback. 
Before answering your question, I am trying to understand where your comments are coming from. Isn't this PR specifically for adding list indexing to advanced indexing? I am not sure why you are talking about "advanced boolean indexing". Are you suggesting the implementation of list indexing in advanced indexing is not needed? Would you able to explain this for me? I would like to better understand your thoughts.

File: thunder/clang/__init__.py
Comment (t-vi) [reply thunder/clang/__init__.py:730]: In order to do the linear indices / reduce to take we would want tensors for tensor arithmetic.
The list or sequence is a bit complicated due to tuple being special in getitem (so `a[[1, 2, 3]]` is the same as `a[((1, 2, 3),)]` but not as `a[(1, 2, 3)]` which is `a[1, 2, 3]`). For writing things, list clearly using a list is *the* correct thing.

File: thunder/clang/__init__.py
Comment (t-vi) [inline thunder/clang/__init__.py:975]: The question I had yesterday could more succinctly be asked: Should here (or likely more reasonably  `result = prims.slice_prim(a, start_indices, end_indices, strides)`) be an `if` to not perform `slice_prim` when we don't have anything to slice. (But it turns out that having this is not what prevents NVFuser to fail.)

File: thunder/clang/__init__.py
Comment (t-vi) [inline thunder/clang/__init__.py:853]: Should this be with an if in case we already have the right permutation?

File: thunder/clang/__init__.py
Comment (t-vi) [inline thunder/clang/__init__.py:863]: Same as above (Should this be with an if in case we already have the right permutation?).

**Review by t-vi — APPROVED** [review]
Hi @k223kim ! Great work!
I just checked that with NVFuser built after https://github.com/NVIDIA/Fuser/pull/2547 , we pass the CI, so I think we wait for the builds to pick that up an then merge.
If you wanted, you might look at guarding the `slice_prim` and transpose with `if`s to only do them if needed (but we can also do this in a follow-up).

File: thunder/clang/__init__.py
Comment (k223kim) [reply thunder/clang/__init__.py:975]: Thank you for the explanation. I got confused last night. You are right. Let me update it so that it can check if we need indexing at all.

File: thunder/clang/__init__.py
Comment (k223kim) [reply thunder/clang/__init__.py]: Yes, thank you! I have updated accordingly.

File: thunder/clang/__init__.py
Comment (mruberry) [reply thunder/clang/__init__.py:734]: The thing I worry about with converting a list into a tensor is that the list could be constexpr (known at compile-time) while we don't really have a notion of tensors with constexpr data (although we could do that, too, I suppose).

Comment (k223kim) [reply thunder/clang/__init__.py:734]: Do you think it'll be helpful to submit an issue regarding that? Would you be able to provide me an example of the scenario that you are particularly concerned about?

Comment (mruberry) [reply thunder/clang/__init__.py:734]: No, that's OK -- it's a pretty niche concern we probably won't need to bother with for a long time

Comment (k223kim) [reply thunder/clang/__init__.py:734]: Ok, cool. Let me resolve this!

**Comment by mruberry** [comment]
> Hi @mruberry ! It's been a while since I got a review from you :) great to see you! 

Nice to work with you, too!

> Thank you for the detailed feedback. Before answering your question, I am trying to understand where your comments are coming from. Isn't this PR specifically for adding list indexing to advanced indexing? I am not sure why you are talking about "advanced boolean indexing". Are you suggesting the implementation of list indexing in advanced indexing is not needed? Would you able to explain this for me? I would like to better understand your thoughts.

Absolutely! Let's talking about indexing! fyi @nikitaved who may add additional info or correct me.

The best official documentation on indexing tensors is probably https://numpy.org/doc/stable/user/basics.indexing.html, and now there's the newer https://data-apis.org/array-api/2022.12/API_specification/indexing.html, but that document doesn't describe indexing with integer tensors. Let's talk about each separately.

The NumPy documentation has the following categories for indexing:

- basic indexing
- advanced integer indexing
- advanced boolean indexing
- combined advanced and basic indexing

Advanced indexing is triggered when one the key is any of:

- a non-tuple sequence object (like a list)
- an integer or boolean array
- a tuple with one or more sequences, integer arrays, or boolean arrays

Advanced boolean indexing (AKA "boolean array indexing") occurs when the key is a boolean array. This is an operation where the output metadata is dependent on the input's tensor data, and since the tensor data is not known at compile-time we cannot know the output metadata at compile-time. Thunder currently doesn't handle these operations (although there are some ideas about how to do so). The Array API documentation even has a note about this (https://data-apis.org/array-api/2022.12/API_specification/indexing.html#boolean-array-indexing). 

However, any key that doesn't include a boolean tensor does have output metadata that can be determined at compile-time. We could support a lot of advanced indexing by writing the meta function for these cases, then calling a special indexing primitive (which the torch executor would execute by calling PyTorch's getitem). One issue with this approach, however, is that PyTorch does not correctly implement NumPy's indexing, and so we'd have to write a meta function that describes PyTorch's indexing behavior, which could be tricky because I'm not sure PyTorch's indexing behavior is documented.

I don't think the NumPy documentation describing indexing is that great, but it does an OK job explaining the core concepts. Personally, I would have created different indexing categories.

Today Thunder supports:

- basic indexing
- advanced indexing with a sequence of 0D or 1D integer tensors (possibly starting with an ellipsis)
- indexing with a sequence that consists of a single number

This PR can focus narrowly on changing the second bullet from 

- advanced indexing with a sequence of 0D or 1D integer tensors (possibly starting with an ellipsis)

to

- advanced indexing with a sequence of 0D or 1D integer tensors or sequences (possibly starting with an ellipsis)

But it looks like this PR is also trying to improve support for combined indexing?

If the PR just wants to extend the current support for lists (which seems totally reasonable), then what about converting sequences to tensors here?

https://github.com/Lightning-AI/lightning-thunder/blob/6dfe7e939a19d1ef5ab259de8709a79f0104fa42/thunder/clang/__init__.py#L725

```
def _seq_to_tensor(x):
  if isinstance(x, Sequence):
    return tensor_from_sequence(x)
  return x

key = (_seq_to_tensor(k) for k in key)
```

Wouldn't that enable support for empty lists and lists that define a tensor with a single dimension?

**Comment by k223kim** [comment]
Wow, thanks so much for the detailed explanation @mruberry ! I'm always learning from your comments.

Before anything, I would like to clarify one thing; although the PR is named to support 'advanced list indexing', it ended up also touching on combinations of basic and advanced due to the repro examples provided in each issue. Here is an example:
```
def fn(input_ids):
   return input_ids[:, [-1, 0]] 

a = torch.randn(5, 5)
jfn = thunder.jit(fn)
jfn(a)
```
Maybe it would have been less confusing if I had separated the PR into two pieces; one for adding support for list indexing and one for combining advanced and basic indexing. What do you think? (Also curious what @t-vi suggests)

Also, I think your suggestion is valid, however, that means that we wouldn't be able to support something like: `a[[1,2,3]]` (cause in this case, `key` would be the list itself), right? Is that also what you are intending?

**Comment by mruberry** [comment]
> Wow, thanks so much for the detailed explanation @mruberry ! I'm always learning from your comments.
> 
> Before anything, I would like to clarify one thing; although the PR is named to support 'advanced list indexing', it ended up also touching on combinations of basic and advanced due to the repro examples provided in each issue. Here is an example:
> 
> ```
> def fn(input_ids):
>    return input_ids[:, [-1, 0]] 
> 
> a = torch.randn(5, 5)
> jfn = thunder.jit(fn)
> jfn(a)
> ```
> 
> Maybe it would have been less confusing if I had separated the PR into two pieces; one for adding support for list indexing and one for combining advanced and basic indexing. What do you think? (Also curious what @t-vi suggests)
> 

I think separating the work into a few PRs would be helpful because indexing is so tricky it can be hard to review multiple changes at once.

> Also, I think your suggestion is valid, however, that means that we wouldn't be able to support something like: `a[[1,2,3]]` (cause in this case, `key` would be the list itself), right? Is that also what you are intending?

I think that could work because the key is a list, which would trigger advanced indexing (the key is a non-tuple sequence), but the logic here would have to change:

https://github.com/Lightning-AI/lightning-thunder/blob/6dfe7e939a19d1ef5ab259de8709a79f0104fa42/thunder/clang/__init__.py#L718

I think we'd have to do something like create a sequence (lst,), just like we create sequences for tensors like (t,)?

**Comment by k223kim** [comment]
Yeah that could definitely work. I wanted to clarify if that was also part of the "advanced indexing". Thanks for the explanation. 

However, I do want to understand why we would want to separate this PR if both list indexing and mixed indexing (combination of basic and advanced) is working..? Is there a particular concern regarding this implementation? Or is it for the sake of clarity?

File: thunder/clang/__init__.py
Comment (t-vi) [inline thunder/clang/__init__.py:750]: If we wanted to improve this, we could swap the if and else to condition on key being a TensorLike.

**Comment by mruberry** [comment]
> Yeah that could definitely work. I wanted to clarify if that was also part of the "advanced indexing". Thanks for the explanation.
> 
> However, I do want to understand why we would want to separate this PR if both list indexing and mixed indexing (combination of basic and advanced) is working..? Is there a particular concern regarding this implementation? Or is it for the sake of clarity?

It would just make it easier to review. I think the PR adding the conversion of lists into tensors is straightforward, can be merged quickly, and has its own clear tests. All of that makes it a good candidate to separate from the more complicated combined indexing.

**Comment by t-vi** [comment]
FWIW, the thing that is blocking this PR is that the new tests seems to trigger https://github.com/NVIDIA/Fuser/pull/2547 and the fix for that is in git but the packages we pull for PyTorch 2.3 have stopped updating 2 months ago.
Happily @xwang233 is looking into that, and hopefully we can resolve it soon, but the holdup is very orthogonal to the PR itself and at 130 lines plus tests, the PR is largish but does not seem to be excessively complicated.

The scoping of this PR has been discussed and is informed by the user needs, which is a mode that has served us well to make progress in the past few weeks.

> I think we'd have to do something like create a sequence (lst,), just like we create sequences for tensors like (t,)?

As far as I can tell we do handle that case, from the code, the tests, and this:

```
import torch, thunder

def fn(x):
    return x[[1, 2]]

jfn = thunder.jit(fn)
torch.testing.assert_(jfn(torch.arange(5)), fn(torch.arange(5)))
```

I mean, if there is something wrong with the patch, we can either fix it here, or preferably in a follow-up. To my mind it is OK, and people can review or not review based on their appetite.

File: thunder/clang/__init__.py
Comment (mruberry) [inline thunder/clang/__init__.py:601]: Part of the reason I'd like to separate this PR more into updates to advanced indexing and then support for combined indexing is that I'd like to better understand updating basic indexing to support the combined indexing. I think a discussion of approaches for combined indexing support would be really interesting.

File: thunder/clang/__init__.py
Comment (mruberry) [inline thunder/clang/__init__.py:607]: My big questions here are:

- does PyTorch's getitem handle those cases correctly? If not, should happen? Should we try to mimic PyTorch's behavior, or mimic NumPy's behavior?
- what combined indexing case or cases does this PR want to support?

See https://github.com/Lightning-AI/lightning-thunder/blob/6dfe7e939a19d1ef5ab259de8709a79f0104fa42/thunder/tests/opinfos.py#L3502 for some existing discrepancies between PyTorch and NumPy.

I think we previously decided (maybe @nikitaved remembers better) that we should implement NumPy's indexing behavior, and we can leave being consistent with PyTorch (or not) for future work.

The NumPy documentation describes two cases of combined indexing here https://numpy.org/doc/stable/user/basics.indexing.html#combining-advanced-and-basic-indexing:

- the advanced indices are separated by a slice, ellipsis, or newaxis
- the advanced indices are all next to each other

Is this PR addressing one or both of those cases?

With basic indexing, should we require that the advanced indexing elements of a key be replaced by ":" before calling the basic indexing operation? That might be more explicit if someone is trying to call it directly, and would avoid this check in basic indexing.

File: thunder/clang/__init__.py
Comment (mruberry) [inline thunder/clang/__init__.py:709]: This is a performance optimization, right? If so, let's make the comment state that a little more clearly.

What if the strides are not equal to 1? Don't we still need take the slice?

File: thunder/clang/__init__.py
Comment (mruberry) [reply thunder/clang/__init__.py:750]: I'm a little confused by the cases here -- maybe we could go through them in a comment? It seems like there should be some way to simplify the case logic.

Also, for this check (as currently written) isn't it any non-tuple sequence that we'd like to identify?

File: thunder/clang/__init__.py
Comment (mruberry) [inline thunder/clang/__init__.py:759]: Wouldn't this check be any for any Sequence, not just list?

File: thunder/tests/opinfos.py
Comment (mruberry) [inline thunder/tests/opinfos.py:3598]: let's also add a couple cases to the basic indexing cases where:

- all slices are from 0 to the length of their dimension (and so trigger the performance optimization) and have stride of 1
- all slices are from 0 to the length of their dimension (and so trigger the performance optimization) and have strides != 1

File: thunder/clang/__init__.py
Comment (k223kim) [reply thunder/clang/__init__.py:607]: To answer  to best of my knowledge, here are my takes:
-  Thunder currently handles both slice and newaxis as basic indices. (see below)
https://github.com/Lightning-AI/lightning-thunder/blob/7d62ae1c7da8fb57e09da3e95ced734e34480400/thunder/clang/__init__.py#L598
https://github.com/Lightning-AI/lightning-thunder/blob/7d62ae1c7da8fb57e09da3e95ced734e34480400/thunder/clang/__init__.py#L649
- Ellipsis are handled in both advanced and basic indexing
- This PR does not necessarily assume that the advanced indices are all next to each other. That is why I had the `transpose` operation. (this allows us to rearrange so that basic indices come first and advanced indices come after; which leads to 'advanced indices are all next to each other')

File: thunder/tests/opinfos.py
Comment (k223kim) [reply thunder/tests/opinfos.py:3598]: Let me add this :) thank you!

File: thunder/clang/__init__.py
Comment (k223kim) [reply thunder/clang/__init__.py:750]: What other sequences do we want to cover in advanced indexing? (Slices are covered in basic indexing)

File: thunder/clang/__init__.py
Comment (k223kim) [reply thunder/clang/__init__.py:709]: You are right. I missed that. Thank you for pointing it out. Let me update this!

File: thunder/clang/__init__.py
Comment (k223kim) [reply thunder/clang/__init__.py:601]: My intention here was to convert any "advanced indices" with slice(None,None,None) so that basic indexing could "skip" it (i.e. no effect to basic indexing). To update that advanced key, I needed a variable to keep those basic keys + converted advanced keys.

File: thunder/clang/__init__.py
Comment (k223kim) [reply thunder/clang/__init__.py:764]: I would like to understand what other Sequence we want to cover in advanced indexing. Would you be able to explain that? Or provide me an example?

**Comment by t-vi** [comment]
The failing test `test_cse_rematerialization` seems tricky because we optimize out one operation and apparently the names are not that stable.
I would suggest updating as follows:

```diff
diff --git a/thunder/tests/test_nvfuser.py b/thunder/tests/test_nvfuser.py
index 7fcf960d..f2437ad9 100644
--- a/thunder/tests/test_nvfuser.py
+++ b/thunder/tests/test_nvfuser.py
@@ -365,10 +365,25 @@ def test_cse_rematerialization(executor, device, _):
     # fusion groups 1 and 6 correspond with the apply_rotary_emb function
     # Nvfuser with recomputation should use precomputed cos and sin values.
     assert len(fusion_bsyms[1].args) == len(fusion_bsyms[6].args)
-    assert fusion_bsyms[1].subsymbols[0].output.name == "freqs_cos"
-    assert fusion_bsyms[1].subsymbols[1].output.name == "freqs_sin"
-    assert fusion_bsyms[6].subsymbols[0].output.name == "freqs_cos"
-    assert fusion_bsyms[6].subsymbols[1].output.name == "freqs_sin"
+
+    # Below, we check that freqs_sin and tos1 (the latter being freqs_cos) are used
+    # in the same operation in both fusions.
+
+    (fusion1_freqs_sin_arg,) = [a for a in fusion_bsyms[1].args if a.name == "freqs_sin"]
+    (fusion1_tos1_arg,) = [a for a in fusion_bsyms[1].args if a.name == "tos1"]
+    (fusion6_freqs_sin_arg,) = [a for a in fusion_bsyms[6].args if a.name == "freqs_sin"]
+    (fusion6_tos1_arg,) = [a for a in fusion_bsyms[6].args if a.name == "tos1"]
+
+    (fusion1_freqs_sin_user,) = [s for s in fusion_bsyms[1].subsymbols if s.args[0] is fusion1_freqs_sin_arg]
+    (fusion6_freqs_sin_user,) = [s for s in fusion_bsyms[6].subsymbols if s.args[0] is fusion6_freqs_sin_arg]
+
+    assert fusion1_freqs_sin_user.sym is fusion6_freqs_sin_user.sym
+    assert fusion1_freqs_sin_user.args[1:] == fusion6_freqs_sin_user.args[1:]
+    (fusion1_tos1_user,) = [s for s in fusion_bsyms[1].subsymbols if s.args[0] is fusion1_tos1_arg]
+    (fusion6_tos1_user,) = [s for s in fusion_bsyms[6].subsymbols if s.args[0] is fusion1_tos1_arg]
+
+    assert fusion1_tos1_user.sym is fusion6_tos1_user.sym
+    assert fusion1_tos1_user.args[1:] == fusion6_tos1_user.args[1:]
 
 
 # Tests that two separated nvFuser regions can be merged when they don't depend
```

File: thunder/clang/__init__.py
Comment (nikitaved) [reply thunder/clang/__init__.py:607]: NumPy is the definition of (advanced)indexing (https://numpy.org/doc/stable/user/basics.indexing.html) without any surprises. PyTorch tries to comply with NumPy. If not today, it will at some point. Sticking to NumPy is robust and future-proof.

File: thunder/clang/__init__.py
Comment (k223kim) [reply thunder/clang/__init__.py:764]: Let me update this in a follow up PR that can handle more advanced indexing!

**Comment by t-vi** [comment]
Great work @k223kim . Thank you for your insights @mruberry @lantiga . At a week of continuous improvements, let's merge this checkpoint and build from there in a follow-up PR.
If we want, we can also take a few minutes in the technical meeting to day to chat about the next steps for the advanced indexing.

File: thunder/clang/__init__.py
Comment (mruberry) [reply thunder/clang/__init__.py:764]: Following the NumPy documentation, I think any Sequence is something that we should attempt to convert to a tensor?

**Pull request merged**
